spec/support/introspector.rb

```diff
@@ -0,0 +1,82 @@
+module ActiveModel
+  ##
+  # Helps to introspect an ActiveModel class with attributes and validators.
+  #
+  # e.g. +introspector = Introspector.new(klass)+
+  #      +introspector.validator_exists?(:case_insensitive_inclusion, in: %w[Y N])+
+  class Introspector
+    attr_reader :klass
+    def initialize(klass)
+      @klass = klass
+    end
+
+    def attributes
+      @attributes ||= instance.instance_variable_get(:@attributes)
+    end
+
+    def field_exists?(field)
+      instance.attributes.key?(field)
+    end
+
+    def validator_exists?(field, validator_arg)
+      also_match_opts = {}
+
+      key = case validator_arg
+            when Symbol
+              "#{validator_arg.to_s.camelize}Validator"
+            when Hash
+              also_match_opts = validator_arg[validator_arg.keys.first]
+              "#{validator_arg.keys.first.to_s.camelize}Validator"
+            else
+              raise ArgumentError, 'validator_arg takes Symbol or Hash only'
+            end
+
+      validator_class = lookup_validator_class(key)
+
+      klass.validators_on(field).find do |validator|
+        validator_exists = validator.class == validator_class && validator.attributes == [field]
+        if also_match_opts.any?
+          validator_exists && validator.options == also_match_opts
+        else
+          validator_exists
+        end
+      end
+    end
+
+    ##
+    # An ActiveModel symbol type, e.g. +:string+
+    def type_of(field)
+      attributes[field].type.type
+    end
+
+    def instance
+      @instance ||= klass.new(empty_entry)
+    end
+
+    private
+
+    ##
+    # Translate a given symbol like :presence or :ingested_date
+    # to its actual validator class.
+    #
+    # Look in root namespace and ActiveModel::Validations namespace and
+    # give up when we've tried the latter
+    def lookup_validator_class(key)
+      begin
+        validator_class = key.constantize
+      rescue NameError
+        if key.match?(/^ActiveModel/)
+          raise ArgumentError, "Not a validator: '#{key.sub('ActiveModel::Validations::', '')}'"
+        end
+
+        key = "ActiveModel::Validations::#{key}"
+        retry
+      end
+      validator_class
+    end
+
+    def empty_entry
+      @empty_entry ||= Struct.new(:data).new(data: {})
```

Comment:
what's the reason for passing in a struct to the ActiveModel instance?

Reply:
This is just for tests; because `EntryData` has overridden the default initializer to take an `entry`, in order to create an instance such that we can inspect generated metadata we are obliged to send in a thing that looks like an `entry` in that it returns a hash-like object from the `#data` method. A `Struct` is the easiest way of achieving this, although we could also use a `double`.

In an ideal world we'd be able to introspect everything we need just from the class definition, but attributes and their types require us to hydrate an instance.